src/materials/ConcreteASREigenstrain.C

```diff
@@ -0,0 +1,323 @@
+                              "tensile_retention_factor compressive_stress_exponent",
+                              "Advanced");
+
+  params.addParam<unsigned int>(
+      "max_its", 30, "Maximum number of iterations for material solution");
+  params.addParam<bool>(
+      "output_iteration_info", false, "Set true to output material iteration information");
+  params.addParam<bool>("output_iteration_info_on_error",
+                        false,
+                        "Set true to output material iteration information when a step fails");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for material iteration");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-20, "Absolute convergence tolerance for material iteration");
+
+  params.addParamNamesToGroup("max_its output_iteration_info output_iteration_info_on_error "
+                              "relative_tolerance absolute_tolerance",
+                              "Advanced");
+
+  params.addClassDescription(
+      "Computes the volumetric expansion eigenstrain due to alkali-silica reaction.");
+
+  return params;
+}
+
+ConcreteASREigenstrain::ConcreteASREigenstrain(const InputParameters & parameters)
+  : ConcreteExpansionEigenstrainBase(parameters, "ASR"),
+    _temperature(coupledValue("temperature")),
+
+    _rh(coupledValue("relative_humidity")),
+    _rh_exponent(getParam<Real>("rh_exponent")),
+
+    _max_vol_expansion(getParam<Real>("max_volumetric_expansion")),
+    _tau_c_T0(getParam<Real>("characteristic_time")),
+    _tau_L_T0(getParam<Real>("latency_time")),
+    _Uc(getParam<Real>("characteristic_activation_energy")),
+    _UL(getParam<Real>("latency_activation_energy")),
+    _ref_temp(getParam<Real>("reference_temperature")),
+    _alpha(getParam<Real>("stress_latency_factor")),
+
+    _gamma_tensile(getParam<Real>("tensile_absorption_factor")),
+    _gamma_residual(getParam<Real>("tensile_retention_factor")),
+    _beta(getParam<Real>("compressive_stress_exponent")),
+
+    _ASR_dependent_tensile_strength(getParam<bool>("ASR_dependent_tensile_strength")),
+    _beta_f(getParam<Real>("residual_tensile_strength_fraction")),
+
+    _max_its(parameters.get<unsigned int>("max_its")),
+    _output_iteration_info(getParam<bool>("output_iteration_info")),
+    _output_iteration_info_on_error(getParam<bool>("output_iteration_info_on_error")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+
+    _ASR_extent(declareProperty<Real>("ASR_extent")),
+    _ASR_extent_old(getMaterialPropertyOld<Real>("ASR_extent"))
+{
+  const MooseEnum temperature_unit = getParam<MooseEnum>("temperature_unit");
+
+  if (temperature_unit == "Celsius")
+    _temp_offset = 273.15;
+  else if (temperature_unit == "Kelvin")
+    _temp_offset = 0.0;
+  else
+    mooseError("Unsupported temperature unit");
```

Comment:
Do you need this check when using Moose Enum?

Reply:
I think it's a good idea.  It guards against adding an option to the enum and not handling it.

Reply:
I'd recommend a couple of tests for these errors then